Documentation update (#45)
ScottishCovidResponse/Contact-Tracing-Model: docs/doc_v1.0/Build_documentation.md

> SE<sub>1</sub>E<sub>2</sub>I<sub>a</sub>I<sub>s</sub>RD model with the

Reviewer: Change typo on line 12 from 'This model using a...' to 'This model uses a...'

> SE<sub>1</sub>E<sub>2</sub>I<sub>a</sub>I<sub>s</sub>RD model with the
> following transitions allowed:

Reviewer: A few things to change. Perhaps:

```suggestion
This model uses a S-E-Asym-Pre-Sym-Sev-R-D model that allows the following transitions:
```

If the above suggestion is used, be sure to remove line 12 of the un-edited version.

<img src="image1.png" alt="Compartment diagram" width="800">

Figure 1. The compartment model

S – Susceptible – the default condition

E – Exposed, but not infectious

Asym – Infected and asymptomatic, will test positive

Pre – Infected and developing symptoms, will test positive

Sym – Infected and symptomatic, will test positive

Sev – Infected and severely symptomatic, will test positive 

> R – Recovered

Reviewer: Add 'D - Dead'

The inputs are contained in the /inputs folder. They are in two formats,
CSV and JSON.

### InputLocations.json

> The first input file is the inputLocations.json file that allows all the other inputs to be configured. 
> This allows the user to have multiple configurations defined. File locations are relative to the input folder.

Reviewer: I think this needs to be a little clearer. Perhaps the following:

```suggestion
The _InputLocations.json_ file allows the user to specify the location of all other inputs files relative to the _/input_ folder. This allows the user to organise multiple simulation configurations.
```

<img src="inputLocations.png" alt="input Locations" width="600">

### Contacts.csv

> The contact data is a CSV file contain network data that shows contacts

Reviewer: ```suggestion
The contact data is expected to be a CSV file containing network data that shows contacts
```

> of the nodes in contact. The weight field is a measure of the duration,

Reviewer: What is a node in the contact network?

> intensity or proximity of the contact. This is dimensionless number and

Reviewer 2: Suggested change:
The contact data is a CSV file containing the temporal network of contacts between individuals in a population.

is used for determining the spread of the infection and is used for
filtering the contacts should a node be alerted. The addition of a label 
field is to enable filtering of contacts based on their label.

<img src="image2.png" alt="Contacts File" width="500">

> 

Reviewer 2: I'm not quite sure what you intend to say in line 60f., something like "This is dimensionless number and is used as threshold for determining the spread of infection, allowing to filter the individual contacts regarding whether they should be alerted or not."?

Also I suggest to change line 61 to "The contacts data contains an additional field 'label' defining where the contact occurred (e.g. school, work, home, ..). This 'label' field can be used to filter contacts, i.e. to dynamically switch them on or off at the beginning of the simulation."

Author: I've made the first change, but I don't think we ended up implementing the label filtering in the end. Unless it happened when I was away.

person, per timestep. So, 0.05 is 5 people in a population of 100 per
day. This value is likely to be much smaller than 0.05.

<img src="image3.png" alt="Contacts File" width="400">

Figure 3. The diseaseSettings.json file

The exposureProbability4UnitContact (expUnitContact) and exposureExponent (exp) values are 

> used to determine the chance of infection for a given contact. 

Reviewer: I think the mathematical implementation may be worth explicitly declaring.
Also the name _exposureProbability4UnitContact_ sounds too similar to the Java unit testing module. Perhaps a more descriptive name should be added at some point.

every 100 women in this example. This data has been taken from Index
Mundi.

<img src="image4.png" alt="exposureProb" width="300">

Figure 4. The population input

### Initial Exposures

> The initial exposures file contains a list of IDs of the people who are in the
> exposed category at t=0. If this list is longer than the initial exposures, it 
> will be truncated. If it is too short, it will be filled randomly. 

Reviewer: ```suggestion
The initial exposures file contains a list of IDs of the individuals who are
initially exposed at the start of the simulation, i.e. at the first time step. If this list is longer than the initial exposures, it will be truncated. If it is too short, it will be filled randomly. 
```

Infected: the number of the population initially infected at t=0. Must
be greater than 0.

Steady state: if there are still active infections when the contact data
has been processed, the steady state flag enables the active cases to
reach a resolution (recovered or dead) with no further random infections
occurring.

(Optional) Seed: Can be specified on the command line or fixed.

![](image5.png)

Figure 5. the run settings file

## AgeData

The age data file contains a list of id and age. This has been added 
to allows the user to use age graduated contact files. Should the file 
be longer than the population, additional input will be ignored. Should 

> it be shorter, the ages will be generated as per the population settings file. 

Reviewer 3: I think it's not really about how long the age file input is... I think the distinction is more to do with whether or not the id exists in the age data file. If the id exists we will use the provided age, otherwise we would generate it as per the population settings file

![](ageData.png)

## Isolation Policies

The isolation policy can be split into 3 sections.

### Virus Policy

This policy determines the probability of a person in a given compartment 
isolating, and for how long they will isolate. These can be modified and 
varied using the distribution times. Also the time in isolation can vary 
between ABSOLUTE or CONTACT_TIME. This means if a person becomes infected 
they will either isolate for the isolation time from either the time of contact 
or the time they are aware of being infected.
![](virusPolicy.png)

### Default Policy

The default policy allows a global baseline to be set. For example, 

> in an older population 10% of the population may shield.

Reviewer 3: We do not have the versatility yet to condition the policy on the individuals age, but definitely can be added in future with the current framework

![](defaultPolicy.png)

### Alert Policy

The alert policy defines the isolation policy for a person in any state. 

> These follow the same rules as the Virus Policy.

Reviewer 3: Probably worth expanding on this and explaining that these policies are dependent on the alert status of individuals (with a list of the alert status options for clarity)

Infected: the number of the population initially infected at t=0. Must
be greater than 0.

> 

Reviewer 2: Are the 'initialExposures' now called 'Infected'? If not I'd change this to the corresponding variable name (or add in brackets 'initialExposures').

Steady state: if there are still active infections when the contact data
has been processed, the steady state flag enables the active cases to
reach a resolution (recovered or dead) with no further random infections

> occurring.

Reviewer: Is this parameter used?

Author: removed

> 

Reviewer 2: Suggested additions (or has this changed in the input file?):
- "timeStepsPerDay": allows to set the timestep (per day) for disease progression e.g. half days ('2') or quarter days ('4').
- "timeStepSpread": set probability for spreading the disease in each chosen timestep, e.g. for half days [0.4,0.6] or quarter days [0.2, 0.2, 0.4, 0.2].

Can we add an example for the command line argument, e.g 'gradle run --args='--overrideInputFolderLocation=input/scenarios/scenario1a_tracinglevel1 --overrideOutputFolderLocation=output0 --seed=3'

Author: I think the gradle command example should be added to the readme

(Optional) Seed: Can be specified on the command line or fixed.

<img src="image5.png" alt="Run Settings Example" width="300">

Figure 5. the run settings file

## AgeData

> 

Reviewer 2: I suggest renaming AgeData to PopulationData, or IDdata, or PersonData as this contains the IDs of the age-structured individuals (i.e. a list of ID's and associated age).

The age data file contains a list of id and age. This has been added 
to allows the user to use age graduated contact files. Should the file 
be longer than the population, additional input will be ignored. Should 
it be shorter, the ages will be generated as per the population settings file. 

<img src="ageData.png" alt="age Data Example" width="100">

## Isolation Policies

> The isolation policy can be split into 3 sections.

Reviewer: ```suggestion
An isolation policy can be split into 3 sections.
```

### Virus Policy

This policy determines the probability of a person in a given compartment 
isolating, and for how long they will isolate. These can be modified and 
varied using the distribution times. Also the time in isolation can vary 
between ABSOLUTE or CONTACT_TIME. This means if a person becomes infected 
they will either isolate for the isolation time from either the time of contact 
or the time they are aware of being infected.

<img src="virusPolicy.png" alt="virus Policy Example" width="400">

### Default Policy

The default policy allows a global baseline to be set. For example, 
10% of the population may be shielding by default.

> 

Reviewer 2: It's not clear to me what the default policy is and where it is set. Also, I would match the example (% shielding) to the picture.

<img src="defaultPolicy.png" alt="default Policy Example" width="500">

### Alert Policy

> The alert policy defines the isolation policy for a person in any state. 

Reviewer: ```suggestion
The alert policy defines the isolation policy for a person in any alert state. 
```

message is printed describing the error and a Configuration Exception is
thrown.

There are three optional command line interfaces:
- **seed:** allows a seed to be input. Defaults to a random seed.
- **overrideInputFolderLocation:** allows a different input folder to be assigned. Defaults to /input.
- **overrideOutputFolderLocation:** allows a different output folder to be assigned. Defaults to /output.

The input files are described in Inputs.

> 

Reviewer 2: Aha, sorry, I had asked for a command line example somewhere above but maybe it fits better here?

may later have some relation to age. A health score of 1 is assumed to
be a very healthy individual, a score of 0 is a very frail individual.

> #### Isolation Compliance

Reviewer 2: For all of the compliance values, it would be good to add where they are set. For instance, I think some of them are set in the population generator.java, if I'm not mistaken.

### Outbreak

> The outbreak class will be handled in much more detail later on. 

Reviewer: If this is true, it's not explicitly clear that it has been covered. However, I may have just missed it.

the top level of how the propagation algorithm works and where it calls
to. These will be broken down in turn in the following sections.

Upon entering the method, a set of infected cases are read from the initial 
exposures file. If this file is shorter than the value specified in 

> RunSettings, they will be filled to the correct length. Likewise, if it is too 

Reviewer: How are they filled?

long, it is truncated to the correct length.

> 

Reviewer 2: Suggested edits, perhaps:
Upon entering the method, a set of initial infections (index cases) are read from the initial exposures file ('initialExposures.csv'). If this file is shorter than the value of initially infecteds specified in RunSettings.json, they will be filled to the required length by randomly infecting individuals in the population.

Reviewer 2: I used 'index cases' but you use 'seed' somewhere below, so perhaps we should stick with seed or leave 'index cases' out.

### Run Contact Data

> The runContactData method within Outbreak.java is where the main calculations in the code are triggered. 

Reviewer: ```suggestion
The runContactData method within Outbreak.java is where the main calculations are triggered. 
```

## Event Runner

The events within the model are handled by a group of lists that contain specific events that can occur in the model. These are:

> 1.	Alert Event
> 2.	Contact Event
> 3.	Infection Event
> 4.	Virus Event

Reviewer: Since this is a numbered list, perhaps the order should reflect the order they are called in the code.

```suggestion
1.	Virus Event
2.	Infection Event
3.	Contact Event
4.	Alert Event
```

Reviewer 2: I agree with Ed on the order...

Each of these events are processed in different ways and can trigger other events to occur at a future time. 

![](RunContactData.png)

### Contact Event

Contact events are set in the contacts.csv file. They involve the interaction between two people at a given time. 
The duration/proximity of the event is measured using the weight parameter. 

When a contact event takes place the two people are assessed based on their Virus Statuses, this allows the model to 
assess if the contact has the potential to spread the infection. One or both of the contacts may be in isolation, 
so the contact may not take place. The isolation policies are expanded upon below. 

![](ContactEvent.png)

### Alert Event

> The alert event processor is used to cycle through the states of disease alert that a member of the population can be in at any one point. 

Reviewer: Is it worth mentioning the other event processor classes in the other sections?

![](image6.png)

> The person has their status set to a corresponding status and the next status is calculated for a time in the future. 

Reviewer: Which status? Alert status or Virus status?

> The result of the test is determined by whether the individual is currently infectious and a dice roll against the accuracy 
> of the test. The number of false positives and negatives is recorded as an output statistic. 

Reviewer: ```suggestion
The result of a test is determined by whether the individual is currently infectious and a dice roll against the accuracy of a test. The number of false positives and negatives is recorded as an output statistic. 
```

![](AlertEvent.png)

### Infection Event

> An infection event is triggered in three ways:

Reviewer: ```suggestion
An infection event can be triggered in three ways:
```

1. The id has been chosen as a seed infection.
2. The id has been randomly chosen to be a seed.

> 3. The id has been in contact with an infectious individual and has subsequently contracted the infection. 

Reviewer: Replace 'id' with 'individual'

Reviewer 2: Just to add to Ed, maybe worthwhile using 'individual (id)'

> Once a SUSCEPTIBLE individual is exposed to the infection, an infection event is triggered which in turn creates a virus event. 

Reviewer: Is this still true given the order of the event processing in EventRunnerImpl.java has been reversed?

> The infection event occurs one time step after the contact. 

Reviewer: Is this still true given the order of the event processing in EventRunnerImpl.java has been reversed?

See my suggested edit under the Event Runner section.

### Virus Event

> A virus event is the progression of the virus through in compartments. The initial infection is not included as this is 

Reviewer: ```suggestion
A virus event is the progression of the virus through it's compartments. The initial infection is not included as this is 
```

an Infection Event. The processing of the event is very similar to the previous sections, with the exception of checking 
for alerts based on the virus status of the individual. This process creates alert events for people who have been in 

> contact with the individual. The individual has a reportingCompliance field that may reduce the chance of an individual

Reviewer: ```suggestion
contact with an infectious individual. The individual has a reportingCompliance field that may reduce the chance of an individual
```

> A virus event is the progression of the virus through in compartments. The initial infection is not included as this is 
> an Infection Event. The processing of the event is very similar to the previous sections, with the exception of checking 
> for alerts based on the virus status of the individual. This process creates alert events for people who have been in 
> contact with the individual. The individual has a reportingCompliance field that may reduce the chance of an individual
> reporting symptoms. The alert checker and tracing policies are covered below.  

Reviewer: ```suggestion
A virus event is the progression of the virus through intscompartments. The initial infection is not included as this is 
an infection event. The processing of the event is very similar to the previous sections, with the exception of checking 
for alerts based on the virus status of an individual. This process creates alert events for people who have been in 
contact with an infectious individual. An individual has a reportingCompliance field that may reduce the chance of reporting symptoms. The alert checker and tracing policies are covered below.  
```

isolationPolicies.json file. These include policies for global, virus status, alert status and a default. 
The default policy will typically be to not isolate. Each policy has a priority value which allows the most relevant policy to be chosen. 

### Global Policies
Global policies are based on the proportion of the population infected. For example, a “stay at home” policy may be put 
into place if more than 10% of the population are infected, whereas a “stay alert” policy may be in place if 
between 5-10% of the population are infected. These values are as an example and can be configured in the input. 

### 	Virus Status Policies
An isolation policy can be defined based on the virus status of an individual. For example, symptomatic individuals 
may be required to stay at home for a defined period. The proportion required to isolate and duration of the isolation can be defined in the input. 

### 	Alert Status Policies
A policy can be defined for any of the alert statuses, much like with a virus status, the proportion required to isolate 
and duration of the isolation can be defined in the input. 

### 	Determining Isolation Policy

> The isolation status of all individuals is stored in a map. These are used to determine if an individual has already 

Reviewer: Is the fact that the data is stored in a map required?
